src/Microsoft.VisualStudio.Mac.LanguageServices.Razor/DefaultFileChangeTracker.cs

```diff
@@ -0,0 +1,130 @@
+
+using System;
+using Microsoft.CodeAnalysis.Razor;
+using Microsoft.VisualStudio.Editor.Razor;
+using MonoDevelop.Core;
+
+namespace Microsoft.VisualStudio.Mac.LanguageServices.Razor
+{
+    internal class DefaultFileChangeTracker : FileChangeTracker
+    {
+        private readonly ForegroundDispatcher _foregroundDispatcher;
+        private readonly string _normalizedFullPath;
+        private bool _listening;
+
+        public override event EventHandler<FileChangeEventArgs> Changed;
+
+        public DefaultFileChangeTracker(
+            string filePath,
+            ForegroundDispatcher foregroundDispatcher)
+        {
+            if (string.IsNullOrEmpty(filePath))
+            {
+                throw new ArgumentException(Resources.ArgumentCannotBeNullOrEmpty, nameof(filePath));
+            }
+
+            if (foregroundDispatcher == null)
+            {
+                throw new ArgumentNullException(nameof(foregroundDispatcher));
+            }
+
+            FilePath = filePath;
+            _normalizedFullPath = NormalizePath(FilePath);
+            _foregroundDispatcher = foregroundDispatcher;
+        }
+
+        public override string FilePath { get; }
+
+        public override void StartListening()
+        {
+            _foregroundDispatcher.AssertForegroundThread();
+
+            if (_listening)
+            {
+                return;
+            }
+
+            AttachToFileServiceEvents();
+
+            _listening = true;
+        }
+
+        public override void StopListening()
+        {
+            _foregroundDispatcher.AssertForegroundThread();
+
+            if (!_listening)
+            {
+                return;
+            }
+
+            DetatchFromFileServiceEvents();
+
+            _listening = false;
+        }
+
+        // Virtual for testing
+        protected virtual void AttachToFileServiceEvents()
+        {
+            FileService.FileChanged += FileService_FileChanged;
+            FileService.FileCreated += FileService_FileCreated;
+            FileService.FileRemoved += FileService_FileRemoved;
+        }
+
+        // Virtual for testing
+        protected virtual void DetatchFromFileServiceEvents()
+        {
+            FileService.FileChanged -= FileService_FileChanged;
+            FileService.FileCreated -= FileService_FileCreated;
+            FileService.FileRemoved -= FileService_FileRemoved;
+        }
+
+        private void FileService_FileChanged(object sender, FileEventArgs args) => HandleFileChangeEvent(FileChangeKind.Changed, args);
+
+        private void FileService_FileCreated(object sender, FileEventArgs args) => HandleFileChangeEvent(FileChangeKind.Added, args);
+
+        private void FileService_FileRemoved(object sender, FileEventArgs args) => HandleFileChangeEvent(FileChangeKind.Removed, args);
+
+        private void HandleFileChangeEvent(FileChangeKind changeKind, FileEventArgs args)
```

Comment:
There's currently an issue with the MonoDevelop.Core binaries we're compiling against which doesn't allow me to unit test other pieces of the class (they explode on instantiation because of signing goodness). Working on getting this resolved offline.

Reply:
I think that's fine for now. It's a static and it will be very hard to test. This class is also pretty trivial. The only thing here that I really care about tests for is this method.